README.md
[![Python Tests](https://github.com/alifanov/mikasa/actions/workflows/main.yml/badge.svg?branch=master&event=push)](https://github.com/alifanov/mikasa/actions/workflows/main.yml) ![GitHub tag (latest SemVer)](https://img.shields.io/github/v/tag/alifanov/mikasa?label=mikasa&sort=semver)
# Mikasa

Simple backtesting tool

## Installation

```python
pip install git+https://github.com/alifanov/mikasa.git
```

## Usage

```python
import logging

import numpy as np
import pandas as pd

from mikasa import BT, DataSeries, SMAIndicator
import ccxt

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)
exchange = ccxt.binance()


# create strategy for backtesting
class SMABT(BT):
    # set up how to process each bar
    def process_bar(self):
        d = self.datas[0]
        if np.isnan(d[0].sma):
            return
        if not self.position:
            if d[-1].close < d[-1].sma:
                if d[0].close > d[0].sma:
                    self.buy(d[0].close, 0.001)
        else:
            if d[-1].close > d[-1].sma:
                if d[0].close < d[0].sma:
                    self.sell(d[0].close)


if __name__ == "__main__":
    # upload and map data from CSV
    ohlcv = exchange.fetch_ohlcv("BTC/USDT", timeframe="1d", limit=1000)

    df = pd.DataFrame(ohlcv, columns=["datetime", "open", "high", "low", "close", "volume"])
    df["datetime"] = pd.to_datetime(df["datetime"], unit="ms")

    # create DataSeries instance
    ds = DataSeries(df, indicators=[SMAIndicator(period=100)])

    # create instance of BT and set params
    bt = SMABT([ds], balance=1000.0)

    # run backtesting
    bt.run()
    logger.info(f"Profit: ${bt.get_profit():.2f}")
    logger.info(f"ROI: ${bt.get_roi()*100:.2f}%")

    # plot data and indicators
    bt.plot()

```

CHANGELOG.md
## 0.4.2 (2023-04-25)

### Fix

- update example for hypersearch

## 0.4.1 (2023-04-25)

### Fix

- markers for orders on plot

## 0.4.0 (2023-04-24)

### Feat

- add hyper search of params support
- add support for trailing stop loss order

### Fix

- tests to pass
- add tests for orders
- use trade percent to trade amount from price
- add test for orders
- refactor orders

## 0.3.0 (2023-04-20)

### Feat

- add trade pct support
- add commission support

## 0.2.0 (2023-04-15)

### Feat

- add macd indicator

## 0.1.1 (2023-04-15)

### Fix

- update tests

## 0.1.0 (2023-04-15)

### Feat

- update executing orders

## 0.0.2 (2023-04-11)

### Fix

- update example.py and fix bugs

## 0.0.1 (2023-04-11)


mikasa/tests/dataset.py
TEST_DATASET = {
    "Datetime": [
        "2017-01-01",
        "2017-01-02",
        "2017-01-03",
    ],
    "Open": [1.0, 2.0, 3.0],
    "High": [
        2.0,
        3.0,
        4.0,
    ],
    "Low": [0.0, 1.3, 2.0],
    "Close": [1.5, 3.0, 3.5],
    "Volume": [
        100,
        200,
        300,
    ],
}


mikasa/tests/test_bt.py
import pandas as pd

from unittest import TestCase
from .dataset import TEST_DATASET
from ..data import DataSeries
from ..bt import BT


class BTTestCase(TestCase):
    def test_bt(self):
        df = pd.DataFrame(TEST_DATASET).rename(
            columns={
                "Datetime": "datetime",
                "Open": "open",
                "High": "high",
                "Low": "low",
                "Close": "close",
            }
        )
        ds = DataSeries(df)

        bt = BT(ds, balance=1.5)
        self.assertEqual(bt.balance, 1.5)

        bt.process_bar()
        self.assertEqual(bt.balance, 1.5)

        bt.buy(ds[0].close, 1.0)

        bt.go()
        bt.process_open_orders()
        self.assertIsNotNone(bt.fund)
        bt.sell(ds[0].close)
        bt.go()
        bt.process_open_orders()
        self.assertEqual(bt.fund, 0)
        self.assertEqual(bt.get_profit(), 1.4985)
        self.assertEqual(bt.get_roi(), 0.999)

        self.assertEqual(bt.dataseries[0].close, 3.5)

    def test_bt_strategy(self):
        class SimpleBT(BT):
            def process_bar(self):
                if self.dataseries[0].close == 1.5:
                    self.buy(self.dataseries[0].close, 1.0)
                if self.dataseries[0].close == 3.0:
                    self.sell(self.dataseries[0].close)

        df = pd.DataFrame(TEST_DATASET).rename(
            columns={
                "Datetime": "datetime",
                "Open": "open",
                "High": "high",
                "Low": "low",
                "Close": "close",
            }
        )
        ds = DataSeries(df)
        bt = SimpleBT(ds, balance=1.5)
        bt.run()

        self.assertEqual(bt.fund, 0)
        self.assertEqual(bt.get_profit(), 1.4985)
        self.assertEqual(bt.get_roi(), 0.999)


mikasa/tests/test_data.py
import pytest

import pandas as pd

from ..data import DataPoint, DataSeries, DataSeriesException
from unittest import TestCase

from .dataset import TEST_DATASET


class DataTestCase(TestCase):
    def test_datapoint(self):
        dp = DataPoint({'foo': 'bar'})
        self.assertEqual(dp.foo, 'bar')

    def test_dataseries(self):
        df = pd.DataFrame(TEST_DATASET).rename(columns={
            'Datetime': 'datetime',
            'Open': 'open',
            'High': 'high',
            'Low': 'low',
            'Close': 'close',
            'Volume': 'volume',
        })
        ds = DataSeries(df)
        self.assertEqual(ds.index, 0)
        self.assertEqual(ds.length, 3)

        self.assertEqual(ds[0].close, 1.5)

        ds.next()
        self.assertEqual(ds.index, 1)

        ds.next()
        self.assertEqual(ds.index, 2)
        self.assertEqual(ds.is_end(), True)

        self.assertEqual(ds[-2].close, 1.5)

        with pytest.raises(DataSeriesException):
            ds[-3]

example.py
import logging

import numpy as np
import pandas as pd

from mikasa import BT, DataSeries, SMAIndicator
import ccxt

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)
exchange = ccxt.binance()


# create strategy for backtesting
class SMABT(BT):
    # set up how to process each bar
    def process_bar(self):
        d = self.dataseries
        if np.isnan(d[0].sma):
            return
        if d[-1].close < d[-1].sma:
            if d[0].close > d[0].sma:
                self.buy(d[0].close, 0.005)
        if d[-1].close > d[-1].sma:
            if d[0].close < d[0].sma:
                self.sell(d[0].close)


if __name__ == "__main__":
    # upload and map data from CSV
    ohlcv = exchange.fetch_ohlcv("BTC/USDT", timeframe="1d", limit=1000)

    df = pd.DataFrame(ohlcv, columns=["datetime", "open", "high", "low", "close", "volume"])
    df["datetime"] = pd.to_datetime(df["datetime"], unit="ms")

    # create DataSeries instance
    ds = DataSeries(df, indicators=[SMAIndicator(period=100)])

    # create instance of BT and set params
    bt = SMABT(ds, balance=1000.0)

    # run backtesting
    bt.run()
    logger.info(f"Profit: ${bt.get_profit():.2f}")
    logger.info(f"ROI: ${bt.get_roi()*100:.2f}%")

    # plot data and indicators
    bt.plot()


mikasa/bt.py
import logging

import matplotlib.pyplot as plt

from mikasa.orders import LimitOrder, OrderType, TrailingStopLossOrder

logger = logging.getLogger(__name__)


class BTException(Exception):
    pass


class BT:
    trade_pct = 0.01
    trade_amount = None
    commission_fraction = 0.001

    def __init__(self, dataseries, balance=1000.0, verbose=True):
        self.dataseries = dataseries
        self.balance = balance
        self.start_balance = balance
        self.fund = 0
        self.open_orders = []
        self.order_history = []
        self.verbose = verbose

    def cancel_open_orders(self):
        for order in self.open_orders:
            if self.verbose:
                logger.info(f"Canceled [{order.type.upper()}]")
        self.open_orders = []

    def get_trade_amount(self, price: float = None):
        if price is None:
            raise ValueError("price is None")
        if self.trade_amount is None:
            return (self.balance * self.trade_pct) * price
        return self.trade_amount

    def prepare_data(self, indicators=None):
        if indicators is not None:
            self.dataseries.add_indicators(indicators=indicators)

    def buy(self, price, shares_volume):
        if self.balance < price * shares_volume:
            if self.verbose:
                logger.warning("Can not execute order due to not enough balance")
            return

        dt = self.dataseries[0].datetime
        order = LimitOrder(type=OrderType.BUY, price=price, volume=shares_volume, executed_at=None)
        self.open_orders.append(order)
        if self.verbose:
            logger.info(f"[BUY] created {dt=} {price=:.2f} {shares_volume=:.2f}")

        trail_percent = 0.2
        trailing_stop_loss_order = TrailingStopLossOrder(
            type=OrderType.SELL,
            price=price * (1.0 - trail_percent),
            volume=shares_volume,
            trail_percent=trail_percent,
            executed_at=None,
        )
        self.open_orders.append(trailing_stop_loss_order)
        if self.verbose:
            logger.info(f"Stop loss [SELL] created {dt=} {price=:.2f} {shares_volume=:.2f}")

    def sell(self, price):
        if self.fund:
            dt = self.dataseries[0].datetime
            order = LimitOrder(type=OrderType.SELL, price=price, volume=self.fund, executed_at=None)
            self.open_orders.append(order)
            if self.verbose:
                logger.info(f"[SELL] created {dt=} {price=:.2f} volume={self.fund:.2f}")

    def process_bar(self):
        pass

    def get_profit(self):
        return self.balance - self.start_balance

    def get_roi(self):
        return 1.0 * self.get_profit() / self.start_balance

    def go(self):
        return self.dataseries.next()

    def process_open_orders(self):
        dt = self.dataseries[0].datetime
        dp = self.dataseries[0]
        rest_orders = []
        for order in self.open_orders:
            if order.can_be_executed(dp.high, dp.low, self.fund, self.balance):
                fund, balance = order.execute(dt=dt, commission_fraction=self.commission_fraction)
                self.order_history.append(order)
                self.fund += fund
                self.balance += balance
                if self.verbose:
                    logger.info(f"Executed [{order.type}] {dt=} balance={self.balance:.2f}")
            else:
                order.update_trailing_state(dp.high, dp.low)
                rest_orders.append(order)
        self.open_orders = rest_orders

    def run(self):
        self.prepare_data()

        while not self.dataseries.is_end():
            self.process_open_orders()
            self.process_bar()
            self.go()

        self.sell(self.dataseries[0].close)

        self.process_open_orders()

        self.cancel_open_orders()

    def plot(self, title):  # pragma: no cover
        data = self.dataseries.data
        data.set_index("datetime", inplace=True)
        headers = ["close"]
        outline_indicators = []
        outline_data = {}
        for ind in self.dataseries.indicators:
            if ind.draw_inline:
                headers.append(ind.title)
            else:
                outline_indicators.append(ind)
                outline_data[ind.title] = data[ind.title]
                data.drop(ind.title, inplace=True, axis=1)

        data = data[headers]
        nrows = len(outline_indicators) + 1
        fig, axes = plt.subplots(nrows=nrows, ncols=1)

        data.plot(ax=axes[0], sharex=True, title=title)

        for i, oi in enumerate(outline_indicators):
            outline_data[oi.title].plot(ax=axes[i + 1], sharex=True, title=oi.title)
            oi.draw_extra_charts(axes[i + 1])

        buy_orders = [it for it in self.order_history if it.type == OrderType.BUY]
        sell_orders = [it for it in self.order_history if it.type == OrderType.SELL]

        self._draw_orders(buy_orders, marker="g^", ax=axes[0])
        self._draw_orders(sell_orders, marker="rv", ax=axes[0])

        plt.show()

    def _draw_orders(self, orders, marker, ax):
        order_datetimes = [order.executed_at for order in orders]
        order_prices = [order.price for order in orders]
        ax.plot(order_datetimes, order_prices, marker)


mikasa/data.py
from stockstats import StockDataFrame


class DataPoint:
    def __init__(self, dt):
        self.dt = dt

    def __getattr__(self, key):
        return self.dt[key]


class DataSeriesException(Exception):
    pass


class DataSeries:
    def __init__(self, data, index=0):
        self.indicators = []
        self.data = StockDataFrame.retype(data.copy())
        self.index = index
        self.prepare()

    def prepare(self):
        data_dict = self.data.to_dict(orient="split")
        self._data = [DataPoint({k: v for k, v in zip(data_dict["columns"], d)}) for d in data_dict["data"]]

    def add_indicators(self, indicators):
        self.indicators = indicators
        for indicator in self.indicators:
            self.data[indicator.title] = indicator.get_data(self.data, "close")
        self.prepare()

    @property
    def length(self):
        return self.data.shape[0]

    def next(self):
        self.index += 1
        return self._data[self.index]

    def __getitem__(self, index):
        if (index + self.index) < 0:
            raise DataSeriesException(
                "Result index is negative. Inner index: {}. Parameter index: {}".format(self.index, index)
            )
        return self._data[index + self.index]

    def is_end(self):
        return self.index == (self.length - 1)


mikasa/orders.py
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Optional


class OrderType(str, Enum):
    BUY = "BUY"
    SELL = "SELL"


@dataclass
class BaseOrder:
    type: OrderType
    price: float
    volume: float
    executed_at: Optional[datetime]

    def _is_enough_balance(self, balance):
        return balance >= self.price * self.volume

    def _is_enough_fund(self, fund):
        return fund >= self.volume

    def update_trailing_state(self, high, low):
        pass

    def can_be_executed(self, high, low, fund, balance):
        if self.type == OrderType.BUY:
            if not self._is_enough_balance(balance):
                return False
            return low < self.price
        if self.type == OrderType.SELL:
            if not self._is_enough_fund(fund):
                return False
            return high > self.price
        return False

    def execute(self, dt, commission_fraction):
        balance_delta = self.volume * self.price * (1.0 - commission_fraction)
        if self.type == OrderType.BUY:
            self.executed_at = dt
            return self.volume, -balance_delta
        if self.type == OrderType.SELL:
            self.executed_at = dt
            return -self.volume, balance_delta
        raise ValueError("Order type is bad")

    class Meta:
        abstract = True


class LimitOrder(BaseOrder):
    pass


class StopOrder(BaseOrder):
    def can_be_executed(self, high, low, balance, fund):
        if self.type == OrderType.BUY:
            if not self._is_enough_balance(balance):
                return False
            return high > self.price
        if self.type == OrderType.SELL:
            if not self._is_enough_fund(fund):
                return False
            return low < self.price
        return False


@dataclass
class TrailingStopLossOrder(StopOrder):
    trail_percent: float

    def update_trailing_state(self, high, low):
        if self.type == OrderType.SELL:
            self.price = low * (1.0 - self.trail_percent)


mikasa/indicators.py
class BaseIndicator:
    def __init__(self, title="", draw_inline=True):
        self.title = title
        self.draw_inline = draw_inline

    def draw_extra_charts(self, *args, **kwargs):  # pragma: no cover
        pass


class SMAIndicator(BaseIndicator):
    def __init__(self, period, title="sma"):
        super(SMAIndicator, self).__init__(title)
        self.period = period

    def get_data(self, df, field_name):
        ds = df[field_name]
        return ds.rolling(center=False, window=self.period).mean()


class MomentumIndicator(BaseIndicator):
    def __init__(self, period=14, title="momentum"):
        super(MomentumIndicator, self).__init__(title)
        self.period = period

    def get_data(self, df, field_name):
        shifted = df[field_name].shift(-self.period)
        return 100.0 * df[field_name] / shifted


class RSIIndicator(BaseIndicator):
    def __init__(self, period=14, title="rsi"):
        super(RSIIndicator, self).__init__(title, draw_inline=False)
        self.period = period

    def draw_extra_charts(self, axe):  # pragma: no cover
        axe.axhline(y=20, xmin=0, xmax=1, c="red", zorder=0, linewidth=1)
        axe.axhline(y=80, xmin=0, xmax=1, c="green", zorder=0, linewidth=1)

    def get_data(self, df, field_name):
        ds = df[field_name]
        delta = ds.diff()
        d_up, d_down = delta.copy(), delta.copy()
        d_up[d_up < 0] = 0
        d_down[d_down > 0] = 0

        rol_up = d_up.rolling(center=False, window=self.period).mean()
        rol_down = d_down.rolling(center=False, window=self.period).mean().abs()

        rs = rol_up / rol_down
        rsi = 100.0 - (100.0 / (1.0 + rs))
        return rsi


class EMAIndicator(BaseIndicator):
    def __init__(self, period=14, title="ema"):
        super(EMAIndicator, self).__init__(title)
        self.period = period

    def get_data(self, df, field_name):
        return df["{}_{}_ema".format(field_name, self.period)]


class MACDIndicator(BaseIndicator):
    def __init__(self, short_period=12, long_period=26, signal_period=9, title="macd"):
        super(MACDIndicator, self).__init__(title, draw_inline=False)
        self.short_period = short_period
        self.long_period = long_period
        self.signal_period = signal_period
        self.draw_inline = False

    def draw_extra_charts(self, axe):  # pragma: no cover
        axe.axhline(y=0, xmin=0, xmax=1, c="red", zorder=0, linewidth=1)

    def get_data(self, df, field_name):
        short_ema = df[field_name].ewm(span=self.short_period, adjust=False).mean()
        long_ema = df[field_name].ewm(span=self.long_period, adjust=False).mean()
        macd = short_ema - long_ema
        signal_line = macd.ewm(span=self.signal_period, adjust=False).mean()

        return macd - signal_line


macd_example.py
import logging

import numpy as np
import pandas as pd

from mikasa import BT, DataSeries, MACDIndicator
import ccxt

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)
exchange = ccxt.binance()


# create strategy for backtesting
class MACDBT(BT):
    trade_amount = 0.005
    indicators = [MACDIndicator()]

    def add_indicators(self, indicators=None):
        self.dataseries.add_indicators(indicators=self.indicators)
        if indicators is not None:
            self.dataseries.add_indicators(indicators=indicators)

    # set up how to process each bar
    def process_bar(self):
        d = self.dataseries
        if np.isnan(d[0].macd):
            return
        if d[0].macd > 0 > d[-1].macd:
            self.buy(d[0].close, self.get_trade_amount(d[0].close))


class HyperSearch:
    result = []

    def get_best_run(self):
        best_run = sorted(self.result, key=lambda x: x[0])[-1]
        return best_run

    def get_params(self):
        return {
            "pair": ["BTC/USDT", "ETH/USDT"],
            "timeperiod": ["5m", "1h", "1d"],
            "macd": {"short_period": range(2, 12), "long_period": range(3, 30), "signal_period": range(4, 16)},
        }

    def run(self):
        params = self.get_params()
        for pair in params["pair"]:
            for timeperiod in params["timeperiod"]:
                ohlcv = exchange.fetch_ohlcv(pair, timeframe=timeperiod, limit=300)

                df = pd.DataFrame(ohlcv, columns=["datetime", "open", "high", "low", "close", "volume"])
                df["datetime"] = pd.to_datetime(df["datetime"], unit="ms")

                macd_params = params["macd"]

                # create DataSeries instance
                for short_period in macd_params["short_period"]:
                    for long_period in macd_params["long_period"]:
                        for signal_period in macd_params["signal_period"]:
                            logger.info(
                                f"Backtesting params: {pair=} {timeperiod=}"
                                f" {short_period=} {long_period=} {signal_period=}"
                            )

                            ds = DataSeries(df)
                            bt = MACDBT(ds, balance=1000.0, verbose=False)
                            bt.add_indicators(
                                indicators=[
                                    MACDIndicator(
                                        short_period=short_period,
                                        long_period=long_period,
                                        signal_period=signal_period,
                                    )
                                ]
                            )

                            # run backtesting
                            bt.run()
                            logger.info(f"Profit: ${bt.get_profit():.2f}")
                            logger.info(f"ROI: ${bt.get_roi() * 100:.2f}%")

                            self.result.append(
                                (
                                    bt.get_roi(),
                                    bt.get_profit(),
                                    {
                                        "pair": pair,
                                        "timeperiod": timeperiod,
                                        "macd": {
                                            "short_period": short_period,
                                            "long_period": long_period,
                                            "signal_period": signal_period,
                                        },
                                    },
                                )
                            )


def backtest_best_run(short_period, long_period, signal_period):
    # upload and map data from CSV
    ohlcv = exchange.fetch_ohlcv(
        best_params["pair"],
        timeframe=best_params["timeperiod"],
        limit=300,
    )

    df = pd.DataFrame(ohlcv, columns=["datetime", "open", "high", "low", "close", "volume"])
    df["datetime"] = pd.to_datetime(df["datetime"], unit="ms")

    # create DataSeries instance
    ds = DataSeries(df)

    bt = MACDBT(ds, balance=1000.0)
    bt.add_indicators(
        indicators=[
            MACDIndicator(
                short_period=short_period,
                long_period=long_period,
                signal_period=signal_period,
            )
        ]
    )

    # run backtesting
    bt.run()
    logger.info(f"Profit: ${bt.get_profit():.2f}")
    logger.info(f"ROI: ${bt.get_roi() * 100:.2f}%")

    # plot data and indicators
    bt.plot(title=f'{best_params["pair"]} ({best_params["timeperiod"]})')


if __name__ == "__main__":
    hyper_search = HyperSearch()
    hyper_search.run()
    best_run = hyper_search.get_best_run()
    logger.info(f"Best run: profit={best_run[1]:.2f}")
    logger.info(f"Best run: ROI={best_run[0] * 100.:.2f}")
    best_params = best_run[2]
    logger.info(f"Best params: {best_run=}")

    macd_best_params = best_params["macd"]
    short_period = macd_best_params["short_period"]
    long_period = macd_best_params["long_period"]
    signal_period = macd_best_params["signal_period"]

    backtest_best_run(short_period, long_period, signal_period)
